### PR TITLE
Added Subscriptions module, Transactions module

### PR DESCRIPTION
**Pull request by Zainab-5**
> Buyers function to make a subscription.
>Admin's function to edit, add features.
>Created Transaction table to store transactions.
>Updated routes.
>Applied ajax to delete features and plans.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,13 +1,21 @@
 class ApplicationController < ActionController::Base
   include Pundit
   protect_from_forgery with: :exception, prepend: true
-  #before_action :authenticate_user!
+  before_action :authenticate_user!, if: :devise_controller?
   before_action :configure_permitted_parameters, if: :devise_controller?
```
Comment (awais-aman) [inline app/controllers/application_controller.rb]: both of these can be in single line.

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,13 +1,21 @@
...
 
   protected
 
   def configure_permitted_parameters
-    devise_parameter_sanitizer.permit(:sign_up) { |u| u.permit(:name, :email, :password, :image, :type) }
+    devise_parameter_sanitizer.permit(:sign_up) { |u| u.permit(:name, :email, :password, :image, :type ) }
   # devise_parameter_sanitizer.permit(:account_update) { |u| u.permit(:name, :email, :password, :current_password) }
```
Comment (awais-aman) [inline app/controllers/application_controller.rb]: remove commented code

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,13 +1,21 @@
...
   end
+
+  def after_sign_in_path_for(resource)
+    if current_user.type == "Admin"
```
Comment (awais-aman) [inline app/controllers/application_controller.rb]: can be a single line `current_user.path == 'admin' ? plans_path : subscriptions_path`

File: app/controllers/features_controller.rb [hunk 4]
```diff
@@ -1,12 +1,45 @@
 class FeaturesController < ApplicationController
   def create
-    @user = User.find(params[:plan_id])
-    @plan = @user.plans.create(feature_params)
-    redirect_to welcome_index_path
+    @feature = Feature.create(feature_params)
+    redirect_to features_path
```
Comment (awais-aman) [inline app/controllers/features_controller.rb]: What if create throws an error

File: app/controllers/features_controller.rb [hunk 5]
```diff
@@ -1,12 +1,45 @@
...
+  end
+
+  def index
+    @feature = Feature.all
```
Comment (awais-aman) [inline app/controllers/features_controller.rb]: variable names should be describing the type it holds. In this case it should be features.

File: app/controllers/features_controller.rb [hunk 6]
```diff
@@ -1,12 +1,45 @@
...
+      if @feature.destroy
+        format.html { redirect_to welcome_index_path, notice: 'feature was successfully created.' }
+        format.js
+      end
+    end
+  end
+
+  def edit
+    @feature = Feature.find(params[:id])
+  end
```
Comment (awais-aman) [inline app/controllers/features_controller.rb:34]: add authorization here who can edit.

File: app/controllers/plans_controller.rb [hunk 7]
```diff
@@ -1,15 +1,50 @@
...
+     @plan = Plan.new
+  end
+
+  def create
+    @user = User.find(current_user.id)
+    @plan = @user.plans.new(plan_params)
+
+    respond_to do |format|
+      if @plan.save
+        format.html { redirect_to plans_path, notice: 'Plan was successfully created.' }
```
Comment (awais-aman) [inline app/controllers/plans_controller.rb]: what if plan doesn't save and has validation errors.

File: app/controllers/plans_controller.rb [hunk 8]
```diff
@@ -1,15 +1,50 @@
...
+  def new
+     @plan = Plan.new
+  end
+
+  def create
+    @user = User.find(current_user.id)
+    @plan = @user.plans.new(plan_params)
+
+    respond_to do |format|
+      if @plan.save
```
Comment (awais-aman) [inline app/controllers/plans_controller.rb]: Use @user.plans.create(plan_params). No need to use @plan = @user.plans.new(plan_params)

File: app/controllers/subscriptions_controller.rb [hunk 9]
```diff
@@ -1,10 +1,26 @@
...
+    redirect_to subscriptions_path
+  end
+
+  def new
+    @subscription = Subscription.new
+  end
+
+  def index
+    if(current_user.type == 'Buyer')
+      @subscriptions = Subscription.where(user_id: current_user.id).all
```
Comment (awais-aman) [inline app/controllers/subscriptions_controller.rb]: Use current_user.subscriptions here

File: app/controllers/usages_controller.rb [hunk 10]
```diff
@@ -0,0 +1,4 @@
+class UsagesController < ApplicationController
+  def index
+  end
```
Comment (awais-aman) [inline app/controllers/usages_controller.rb:6]: single line method

**Review by awais-aman — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (Zainab-5) [reply app/controllers/application_controller.rb]: Done.

File: app/controllers/features_controller.rb (hunk 4 above)
Comment (Zainab-5) [reply app/controllers/features_controller.rb]: Handled.

File: app/controllers/features_controller.rb (hunk 5 above)
Comment (Zainab-5) [reply app/controllers/features_controller.rb]: Corrected.

File: app/controllers/features_controller.rb (hunk 6 above)
Comment (Zainab-5) [reply app/controllers/features_controller.rb:34]: Done.

File: app/controllers/usages_controller.rb [hunk 11]
```diff
@@ -1,22 +1,17 @@
 class UsagesController < ApplicationController
   def create
-    @user = User.find(current_user.id)
-    @usage =  @user.transactions.create(transaction_params)
+    @usage = Usage.create(units_used: params[:usage][:units_used], subscription_id: params[:subscription_id], feature_id: params[:feature_id])
+    byebug
```
Comment (Zainab-5) [inline app/controllers/usages_controller.rb:4]: Byebug removed.

File: app/controllers/usages_controller.rb (hunk 10 above)
Comment (Zainab-5) [reply app/controllers/usages_controller.rb:6]: Corrected.

**Pull request merged**